docs/job-writing-guide.md

```diff
@@ -0,0 +1,750 @@
+allows the runtime to lazily evaluate the value at that path - see Reading state
+lazily. Not all operations support a JSON path string - refer to the docs. :::
+
+## Variable initialisation
+
+<!--
+  I'm a bit iffy on this because actually in v2 you can just do const result = [] at the top of your job
+
+  Is the fn block really better practice?
+-->
+
+A common pattern is to need to declare some variables at the state of the job.
+These could be static values to use later, functions to be called multiple times
+through the job, or bits of state that we want to return at the end.
+
+It is considered best practice to use an `fn()` block to do this at the start of
+the job, writing all values to state.
+
+```js
+fn(state => {
+  // Create an array to hold the final results of the job
+  state.results = [];
+
+  // Create a lookup index to be used during the job
+  state.lookup = {};
+
+  state.keyMap = {
+    AccountName: 'C__Acc_Name', // capture various static mappings for transformation
+  };
+
+  state.maxPageSize = 200; // Define some config options
+
+  state.convertToSF = item => {
+    /* ... */
+  }; // A function to be re-used
+
+  return state;
+});
+
+// the rest of your job code goes here
+get('/abc');
+
+fn(state => {
+  /* ... */
+
+  // Only return the results array as output from the job
+  return { result: state.results };
+});
+```
+
+## Cleaning final state
+
+When your job has completed, the final state object will be "returned" by the
+openfn runtime.
+
+This final state must be serialisable to JSON.
+
+If there are more steps in the workflow, the state will be passed to those
+steps. If running on the app (Lightning), the final state will be saved as a
+dataclip. Or if running in the CLI, the final state will be written to disk.
+
+It's often desirable to clean up your final state so that any unused information
+is removed. This reduces the size of your saved data, but could also be an
+important security consideration.
+
+The best way to do this is with a closing `fn()` block which returns just the
+keys you want (this is usually best):
+
+```js
+fn(state => {
+  return {
+    data: state.data,
+  };
+});
+```
+
+You could use the spread operator to override some keys:
+
+```js
+fn(state => {
+  return {
+    ...state,
+    secretStuff: null,
+  };
+});
+```
+
+Or use the rest operator:
+
+```js
+fn(state => {
+  const { usename, password, secrets, ...rest } = state;
+  return rest;
+});
+```
+
+:::info Configuration & Functions OpenFn will automatically scrib the
+`configuration` key and any functions from your final state. :::
+
+<!--
+I would like to include this BUT fields is not an operation and so works a bit differently
+
+## Object assembly with fields()
+
+Another common pattern is to have to transform one object to another.
+
+The `fields()` operator is a conveient way to do this. `fields()` will assemble an object based on the keys and values you provide. It composes well witth other operations.
+-->
+
```

Comment:
@josephjclark at the end of this section on state, maybe we can link to [this page on initial & final state](https://docs.openfn.org/documentation/build/steps/each)?